File: extractors/go/calls.go
// calls.go — CALLS resolution ladder.
//
// 1. types.Info.Selections[selExpr] → *types.Func → local package → exact
//    METHOD/FUNCTION node; std → skip; external → lazy [mod]/pkg::member
//    (gated), receiver package from the selection's receiver type.
// 2. types.Info.Uses[ident] → *types.Func → same by name.
// 3. Name-based fallback (no type info / external receivers): plain name →
//    same-package func; alias.member → symbol maps → local method/func or
//    [mod] lazy member (python's proven chain rule).
//
// metadata.calls (sorted unique strings) + metadata.resolvedCalls written on
// every caller node (contract compliance + LLM context).

package main

import (
	"go/ast"
	"go/token"
	"go/types"
	"sort"
	"strings"
)

type resolvedCall struct {
	Target string `json:"target"`
	Method string `json:"method,omitempty"`
	Class  string `json:"class,omitempty"`
}

// detectCalls — CALLS edges for every non-test function/method body.
func detectCalls(pr *parsedRepo, ti *TypeInfo, l *LookupMaps, opts *Options) []map[string]any {
	edges := []map[string]any{}

	for _, pf := range pr.files {
		info := ti.Infos[pf.RelPath]
		for _, fn := range pf.Funcs {
			if fn.IsTest {
				continue // test funcs are leaf metadata, not callers
			}
			callerID := nodeIDForFunc(pf.RelPath, fn)
			callStrings := make([]string, 0, len(fn.Calls))
			var details []resolvedCall
			for _, c := range fn.Calls {
				callStrings = append(callStrings, c.Str)
				target, method, class, e := resolveCall(c, fn, pf, info, l)
				if e != nil {
					edges = append(edges, e)
				}
				if target != "" {
					details = append(details, resolvedCall{Target: target, Method: method, Class: class})
				}
			}
			// metadata enrichment on the (already created) caller node
			if n, ok := l.NodeByID[callerID]; ok {
				m := n["metadata"].(map[string]any)
				m["calls"] = sortedUnique(callStrings)
				if len(details) > 0 {
					m["resolvedCalls"] = details
				}
			}
		}
	}
	return edges
}

// nodeIDForFunc — matches the id scheme in lookup.go.
func nodeIDForFunc(rel string, fn *ParsedFunc) string {
	if fn.IsMethod {
		return methodNodeID(rel, fn)
	}
	return funcNodeID(rel, fn.Name)
}

// resolveCall → (target node id, method, class, edge) — edge nil when
// unresolvable (documented: builtins/locals/dynamic stay metadata-only).
func resolveCall(c CallSite, fn *ParsedFunc, pf *ParsedFile, info *types.Info,
	l *LookupMaps) (string, string, string, map[string]any) {

	pkgPath := fn.PkgPath
	callerID := nodeIDForFunc(pf.RelPath, fn)

	// ── 1. type-info resolution (exact) ──
	if info != nil {
		if selExpr, ok := c.Node.(*ast.SelectorExpr); ok {
			if sel, ok := info.Selections[selExpr]; ok {
				if obj, ok := sel.Obj().(*types.Func); ok && obj != nil {
					if id, class, ok := funcToNode(obj, l); ok {
						return id, c.Sel, class, callEdge(callerID, id, c.Sel, class)
					}
					if obj.Pkg() != nil {
						path := obj.Pkg().Path()
						if !isStdImport(path) {
							if n := l.tp.memberNode(path, c.Sel); n != nil {
								return n["id"].(string), c.Sel, "", callEdge(callerID, n["id"].(string), c.Sel, "")
							}
						}
					}
				}
			}
		} else if idExpr, ok := c.Node.(*ast.Ident); ok {
			if obj, ok := info.Uses[idExpr].(*types.Func); ok && obj != nil {
				if id, class, ok := funcToNode(obj, l); ok {
					return id, "", class, callEdge(callerID, id, "", class)
				}
				if obj.Pkg() != nil {
					path := obj.Pkg().Path()
					if !isStdImport(path) {
						if n := l.tp.memberNode(path, obj.Name()); n != nil {
							return n["id"].(string), obj.Name(), "", callEdge(callerID, n["id"].(string), obj.Name(), "")
						}
					}
				}
			}
		}
	}

	// ── 1.5 receiver-type-from-declaration (params / typed vars) ──
	// Handlers receive framework types by signature: func(c *gin.Context) —
	// the checker can't resolve external types, but the DECLARED type can.
	if c.IsSel {
		if impPath, typeName, local := receiverDeclaredType(c, fn, pf, l); typeName != "" {
			if local {
				if id, ok := l.MethodNodeByPkgTypeName[impPath+"::"+typeName+"."+c.Sel]; ok {
					return id, c.Sel, typeName, callEdge(callerID, id, c.Sel, typeName)
				}
			} else if n := l.tp.memberNode(impPath, typeName+"."+c.Sel); n != nil {
				return n["id"].(string), c.Sel, typeName, callEdge(callerID, n["id"].(string), c.Sel, typeName)
			}
		}
	}

	// ── 2. name-based fallback ──
	if c.IsSel {
		// alias.member: receiver is an import alias → local member or [mod]
		alias := firstSegment(c.Receiver)
		if impPath, ok := l.SymbolMaps[pf.RelPath][alias]; ok {
			if _, local := l.PkgNodesByPath[impPath]; local {
				if id, ok := l.MethodNodeByPkgTypeName[impPath+"::"+c.Sel]; ok {
					return id, c.Sel, "", callEdge(callerID, id, c.Sel, "")
				}
				if id, ok := l.FuncNodeByPkgName[impPath+"::"+c.Sel]; ok {
					return id, c.Sel, "", callEdge(callerID, id, c.Sel, "")
				}
				return "", "", "", nil
			}
			if n := l.tp.memberNode(impPath, c.Sel); n != nil {
				return n["id"].(string), c.Sel, "", callEdge(callerID, n["id"].(string), c.Sel, "")
			}
			return "", "", "", nil
		}
		// receiver names a local TYPE (User.Create) → method on that type
		if id, ok := l.MethodNodeByPkgTypeName[pkgPath+"::"+c.Receiver+"."+c.Sel]; ok {
			return id, c.Sel, c.Receiver, callEdge(callerID, id, c.Sel, c.Receiver)
		}
		// receiver is a plain variable whose type we can't see — unresolvable
		// tier (documented; type info covers the exact cases)
		return "", "", "", nil
	}

	// plain name: same package first
	if id, ok := l.FuncNodeByPkgName[pkgPath+"::"+c.Str]; ok {
		return id, "", "", callEdge(callerID, id, "", "")
	}
	return "", "", "", nil
}

func callEdge(from, to, method, class string) map[string]any {
	meta := map[string]any{}
	if method != "" {
		meta["method"] = method
	}
	if class != "" {
		meta["class"] = class
	}
	return edgeWithMeta(from, to, EdgeCalls, meta)
}

// receiverDeclaredType — the receiver's type from its DECLARATION: function
// parameters first (func(c *gin.Context) — the documented handler signature),
// then package-level typed vars (var DB *gorm.DB). Returns (importPath,
// typeName, local) — local=false means external (memberNode candidate).
func receiverDeclaredType(c CallSite, fn *ParsedFunc, pf *ParsedFile, l *LookupMaps) (string, string, bool) {
	recvName := firstSegment(c.Receiver)
	if recvName == "" {
		return "", "", false
	}
	// 1. function parameters
	if fn.Decl != nil && fn.Decl.Type != nil && fn.Decl.Type.Params != nil {
		for _, field := range fn.Decl.Type.Params.List {
			for _, name := range field.Names {
				if name.Name == recvName {
					return declaredTypeInfo(field.Type, fn.PkgPath, pf, l)
				}
			}
		}
	}
	// 2. package-level typed vars in this file (var DB *gorm.DB)
	for _, decl := range pf.AST.Decls {
		gd, ok := decl.(*ast.GenDecl)
		if !ok || gd.Tok != token.VAR {
			continue
		}
		for _, spec := range gd.Specs {
			vs, ok := spec.(*ast.ValueSpec)
			if !ok || vs.Type == nil {
				continue
			}
			for _, name := range vs.Names {
				if name.Name == recvName {
					return declaredTypeInfo(vs.Type, fn.PkgPath, pf, l)
				}
			}
		}
	}
	return "", "", false
}

// declaredTypeInfo — a declared type expr → (importPath, typeName, local).
// Ident "User" → local; SelectorExpr "gin.Context" → alias → external path.
func declaredTypeInfo(expr ast.Expr, pkgPath string, pf *ParsedFile, l *LookupMaps) (string, string, bool) {
	switch t := expr.(type) {
	case *ast.StarExpr:
		return declaredTypeInfo(t.X, pkgPath, pf, l)
	case *ast.Ident:
		return pkgPath, t.Name, true // same package — local type
	case *ast.SelectorExpr:
		alias := identName(t.X)
		if impPath, ok := l.SymbolMaps[pf.RelPath][alias]; ok {
			return impPath, t.Sel.Name, false
		}
	case *ast.IndexExpr:
		return declaredTypeInfo(t.X, pkgPath, pf, l)
	case *ast.IndexListExpr:
		return declaredTypeInfo(t.X, pkgPath, pf, l)
	}
	return "", "", false
}

// funcToNode — *types.Func declared in a local package → (node id, typeName).
func funcToNode(obj *types.Func, l *LookupMaps) (string, string, bool) {
	if obj.Pkg() == nil {
		return "", "", false
	}
	pkgPath := obj.Pkg().Path()
	if _, local := l.PkgNodesByPath[pkgPath]; !local {
		return "", "", false
	}
	if recv := obj.Type().(*types.Signature).Recv(); recv != nil {
		typeName := recvTypeNameFromTypes(recv.Type())
		if typeName == "" {
			return "", "", false
		}
		id, ok := l.MethodNodeByPkgTypeName[pkgPath+"::"+typeName+"."+obj.Name()]
		return id, typeName, ok
	}
	id, ok := l.FuncNodeByPkgName[pkgPath+"::"+obj.Name()]
	return id, "", ok
}

// recvTypeNameFromTypes — receiver type → simple name (strips * and pkg path).
func recvTypeNameFromTypes(t types.Type) string {
	switch tt := t.(type) {
	case *types.Pointer:
		return recvTypeNameFromTypes(tt.Elem())
	case *types.Named:
		return tt.Obj().Name()
	}
	return ""
}

func firstSegment(s string) string {
	if i := strings.Index(s, "."); i >= 0 {
		return s[:i]
	}
	return s
}

func sortedUnique(list []string) []string {
	seen := map[string]bool{}
	out := []string{}
	for _, s := range list {
		if !seen[s] {
			seen[s] = true
			out = append(out, s)
		}
	}
	sort.Strings(out)
	return out
}

File: extractors/rust/src/calls.rs
// calls.rs — the CALLS ladder (syntactic — syn does no type resolution).
//
// Ladder (each rung documented; later rungs flagged heuristic where fuzzy):
//   1. Path call `a::b::c()` / `Type::m()`:
//        a. first segment via the file's use-symbol map (alias → target)
//        b. module_map.resolve → (file, rest)
//        c. rest = single name   → FUNCTION in that file (closest_by_path)
//        d. rest = Type::method  → STRUCT/ENUM/TRAIT in that file, then
//           methods_by_owner (same file preferred) → METHOD
//   2. Plain `bar()` → same-file FUNCTION first, then crate-wide name
//   3. `Type::new()` with the type in scope but NOT imported (same module /
//      no use statement) → type via closest_by_path, then its methods
//   4. `self.method()` → methods of the enclosing impl's owner type
//   5. `x.method()` (other receivers) → no type info → metadata only V1
//   6. External first segment (a Cargo.toml dep, not std/local) → lazy
//      [crate]/name::member node (gated)
//
// metadata.calls already carries the raw strings (parser); resolved targets
// are written back as metadata.resolvedCalls.

use crate::contract::{edge, CodeEdge};
use crate::extractor::{Options, ParsedRepo};
use crate::fingerprint::ParsedManifest;
use crate::lookup::LookupMaps;
use crate::module_map::ModuleMap;
use crate::parser::ParsedItem;
use crate::thirdparty::ThirdPartyRegistry;

pub struct CallsOutput {
    pub edges: Vec<CodeEdge>,
}

pub fn detect_calls(
    repo: &ParsedRepo,
    module_map: &ModuleMap,
    lookup: &mut LookupMaps,
    mf: &ParsedManifest,
    tp: &mut ThirdPartyRegistry,
    _opts: &Options,
) -> CallsOutput {
    let mut edges = vec![];

    for pf in &repo.files {
        if pf.is_test_file {
            continue;
        }
        for item in &pf.items {
            if item.is_test || item.calls.is_empty() {
                continue;
            }
            let from = crate::lookup::node_id_for_item(&pf.rel_path, item);
            if from.is_empty() {
                continue;
            }
            let base_module = crate::walker::module_path_for_file(&pf.rel_path);
            let mut resolved: Vec<String> = vec![];
            for call in &item.calls {
                if let Some(target) = resolve_call(
                    call,
                    item,
                    &pf.rel_path,
                    &base_module,
                    module_map,
                    lookup,
                    mf,
                    tp,
                ) {
                    edges.push(edge(&from, &target, "CALLS"));
                    resolved.push(target);
                }
            }
            if !resolved.is_empty() {
                // write metadata.resolvedCalls back (nodes are owned by the
                // registry; clone → mutate → reinsert)
                if let Some(node) = lookup.node_by_id.get(&from).cloned() {
                    let mut n = node;
                    let r: Vec<serde_json::Value> = resolved
                        .iter()
                        .map(|s| serde_json::Value::from(s.clone()))
                        .collect();
                    n.metadata
                        .insert("resolvedCalls".to_string(), serde_json::Value::Array(r));
                    lookup.node_by_id.insert(from, n);
                }
            }
        }
    }

    CallsOutput { edges }
}

/// Resolve one call string → target node id (or None).
#[allow(clippy::too_many_arguments)]
fn resolve_call(
    call: &str,
    item: &ParsedItem,
    rel_path: &str,
    base_module: &str,
    module_map: &ModuleMap,
    lookup: &LookupMaps,
    mf: &ParsedManifest,
    tp: &mut ThirdPartyRegistry,
) -> Option<String> {
    // `x.method()` — receiver form
    if let Some((recv, method)) = call.split_once('.') {
        if recv == "self" {
            // methods of the enclosing impl's owner type
            let owner = &item.impl_target;
            if owner.is_empty() {
                return None;
            }
            return method_of_owner(owner, method, rel_path, lookup);
        }
        // other receivers: no type resolution in V1 — metadata only
        return None;
    }
    resolve_path_call(call, rel_path, base_module, module_map, lookup, mf, tp)
}

/// Resolve a path-form call (`a::b::c`, `Type::m`, plain `bar`) to a node id.
/// Shared by the CALLS ladder and utoipa-axum route handlers (routes.rs).
///
/// Ladder (documented Rust semantics): first segment via the file's use-symbol
/// map → module map resolve (relative-first for call paths) → glob-import
/// unroll (`use crate::controllers::*` brings `krate` into scope) → external
/// crate → lazy [crate]/name::member (gated).
pub fn resolve_path_call(
    call: &str,
    rel_path: &str,
    base_module: &str,
    module_map: &ModuleMap,
    lookup: &LookupMaps,
    mf: &ParsedManifest,
    tp: &mut ThirdPartyRegistry,
) -> Option<String> {
    let segs: Vec<&str> = call.split("::").collect();
    if segs.is_empty() {
        return None;
    }
    if segs.len() == 1 {
        // plain name → same-file FUNCTION, then crate-wide
        if let Some(id) = lookup.closest_by_path(segs[0], rel_path) {
            return Some(id);
        }
        return None;
    }

    // multi-segment path — first segment via symbol map, then module resolve
    let sym_target = lookup
        .symbol_maps
        .get(rel_path)
        .and_then(|syms| syms.get(segs[0]))
        .cloned();
    let full_path = match &sym_target {
        Some(t) => format!("{}::{}", t, segs[1..].join("::")),
        None => call.to_string(),
    };

    // Relative-first module resolution (the call-path ladder). NOTE: when
    // the caller's own file IS a module (e.g. src/router.rs → module
    // "router"), bare paths like `krate::search::list_crates` resolve
    // relative-first to `router::krate::...` → matches the file's OWN module
    // → returns (src/router.rs, "krate::search::list_crates") →
    // resolve_in_file finds no item_by_module_name["router::krate"] → None.
    // That None is NOT a hard failure — fall through to glob unroll +
    // crate-root absolute, where utoipa `routes!(krate::search::list_crates)`
    // handlers actually live (brought into scope via `use crate::controllers::*;`).
    if let Some((file, rest)) = module_map.resolve(&full_path, base_module) {
        if let Some(id) = resolve_in_file(&file, &rest, rel_path, lookup) {
            return Some(id);
        }
        // don't return None — fall through to glob unroll + crate-root
    }

    // glob unroll: `use crate::controllers::*;` brings `krate` into scope, so
    // `krate::search::list_crates` resolves as `crate::controllers::krate::
    // search::list_crates` (documented Rust glob-import semantics)
    if sym_target.is_none() {
        if let Some(globs) = lookup.glob_maps.get(rel_path) {
            for g in globs {
                let cand = format!("{}::{}", g, full_path);
                if let Some((file, rest)) = module_map.resolve(&cand, base_module) {
                    if let Some(id) = resolve_in_file(&file, &rest, rel_path, lookup) {
                        return Some(id);
                    }
                }
            }
        }
    }

    // crate-root absolute fallback. A bare multi-segment path whose first
    // segment is a local module resolves from the CRATE ROOT in edition 2018
    // (same semantics as use-paths — `krate::search::list_crates` from
    // src/router.rs means `crate::krate::search::list_crates`, NOT
    // `router::krate::...`). resolve_use() is crate-root-absolute and won't
    // misresolve against the caller's own module.
    if let Some((file, rest)) = module_map.resolve_use(&full_path, base_module) {
        if let Some(id) = resolve_in_file(&file, &rest, rel_path, lookup) {
            return Some(id);
        }
    }

    // external crate (a dep, not std, not local) → lazy member node
    let first = crate::thirdparty::first_segment(&full_path);
    if crate::thirdparty::is_std(first) {
        return None;
    }
    if crate::thirdparty::is_external_crate(first, mf) {
        let member = full_path.split("::").skip(1).collect::<Vec<_>>().join("::");
        return tp.member_node(first, &member).map(|n| n.id.clone());
    }
    None
}

/// Resolve `rest` (item path within `file`) → node id.
fn resolve_in_file(file: &str, rest: &str, from_rel: &str, lookup: &LookupMaps) -> Option<String> {
    if rest.is_empty() {
        return None;
    }
    let file_module = crate::walker::module_path_for_file(file);
    let parts: Vec<&str> = rest.split("::").collect();
    if parts.len() == 1 {
        // single name — FUNCTION (or struct constructor call `Type(...)`)
        let key = if file_module.is_empty() {
            parts[0].to_string()
        } else {
            format!("{}::{}", file_module, parts[0])
        };
        if let Some(id) = lookup.item_by_module_name.get(&key) {
            return Some(id.clone());
        }
        return lookup.closest_by_path(parts[0], file);
    }
    // Type::method
    let ty_key = if file_module.is_empty() {
        parts[0].to_string()
    } else {
        format!("{}::{}", file_module, parts[0])
    };
    let type_id = lookup.item_by_module_name.get(&ty_key).cloned();
    let method = parts[1];
    if let Some(tid) = type_id {
        if let Some(mid) = method_of_owner_for_type(parts[0], method, file, lookup) {
            let _ = tid;
            return Some(mid);
        }
        // trait-dispatch call: `Trait::method(&x)` — the trait name resolved;
        // find the method among `impl Trait for ...` blocks
        if let Some(mid) = method_of_trait(parts[0], method, lookup) {
            return Some(mid);
        }
        return None;
    }
    // type not found as item — maybe a trait in that module
    let _ = from_rel;
    lookup
        .item_by_module_name
        .get(&format!("{}::{}", file_module, parts[0]))
        .cloned()
        .and_then(|_| method_of_owner(parts[0], method, file, lookup))
}

/// Method of a trait (resolved via `impl Trait for Type` blocks).
fn method_of_trait(trait_name: &str, method: &str, lookup: &LookupMaps) -> Option<String> {
    let ids = lookup.methods_by_trait.get(trait_name)?;
    // id shapes: `rel::Owner.method` (inherent) / `rel::Owner.method::Trait`
    ids.iter()
        .find(|id| id.ends_with(&format!(".{}", method)) || id.contains(&format!(".{}::", method)))
        .cloned()
}

/// Method of an owner type: same-file preferred, then any file.
fn method_of_owner(owner: &str, method: &str, file: &str, lookup: &LookupMaps) -> Option<String> {
    let ids = lookup.methods_by_owner.get(owner)?;
    // id shapes: `rel::Owner.method` / `rel::Owner.method::Trait`
    let matches = |id: &String| {
        id.ends_with(&format!(".{}", method)) || id.contains(&format!(".{}::", method))
    };
    let same_file = ids
        .iter()
        .find(|id| id.starts_with(&format!("{}::", file)) && matches(id));
    match same_file {
        Some(id) => Some(id.clone()),
        None => ids.iter().find(|id| matches(id)).cloned(),
    }
}

fn method_of_owner_for_type(
    owner: &str,
    method: &str,
    file: &str,
    lookup: &LookupMaps,
) -> Option<String> {
    method_of_owner(owner, method, file, lookup)
}
